### PR TITLE
 Provide a higher level implementation for REST handlers 

### LINKED ISSUE #128: [FEATURE] Provide a higher level implementation for REST handlers
### Is your feature request related to a problem?

In https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/sample/rest/RestHelloAction.java#L37 the developer has to implement a `handleRequest` that acts as a router that checks for methods, e.g.

```
if (Method.GET.equals(method) && "/hello".equals(uri)) {
            return new BytesRestResponse(OK, GREETING);
        }
```

This is unnecessarily verbose.

### What solution would you like?

Provide a higher level implementation that implements `ExtensionRestHandler `. As a developer I would like to write something shorter:

```java
public class RestHelloAction extends GenericExtensionRestHandler {

    @Override
    public List<Route> routes() {
        return singletonList(
            new RouteHandler(GET, "/hello", this.handleHello),
            new RouteHandler(GET, "/world", this.handleWorld)
        );
    }

    public RestResponse handleHello(...) {
    }

    public RestResponse handleWorld(...) {
    }
}
```

### LINKED ISSUE #245: [FEATURE] Create boilerplate "not configured to handle" ExtensionRestResponse
### Is your feature request related to a problem?

This code is repeated in every Rest Handler:
```java
return new ExtensionRestResponse(
    request,
    NOT_FOUND,
    "Extension REST action improperly configured to handle " + request.toString()
);
```

### What solution would you like?

Create a method somewhere (`ExtensionRestHandler` default method? Static method on `ExtensionRestResponse`?) that takes the `request` as an argument and returns the result of the above instantiation.

### What alternatives have you considered?

Copying and pasting the above in every Rest Handler implementation.

### Do you have any additional context?

image

### PR DESCRIPTION
**Pull request by dbwiddis**
Signed-off-by: Daniel Widdis <widdis@gmail.com>

### Description

Creates a `RouteHandler` class extending `Route` which also includes a handler method

Creates a `BaseExtensionRestHandler` abstract class:
 - Matches the request's route to its handler
 - Moves the unhandled request boilerplate into the superclass

Note to reviewers: consider [this comment](https://github.com/opensearch-project/opensearch-sdk-java/issues/128#issuecomment-1312656871) for two possible implementations. Happy to switch back to the other one if you don't like the one I chose.  The actual implementation permits either choice; the question is more about what we want the HelloWorld sample to show.

### Issues Resolved

Fixes #128
Fixes #245

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
@mloufra FYI this can simplify your handlers in #213 and #221 if merged before they are (optional to include).

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:24]: Just trying to understand why we need a separate `BaseExtensionRestHandler` class when `ExtensionRestHandler ` can handle the same? Can we move the code there?

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:66]: What's the significance of the 2nd case here?

File: src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java:63]: Looks clean :)

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:66]: the two cases are:
 - handler string exactly matches request string
 - handler string is a named param, such as `{detector_id}`, which matches any string.

File: src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/sample/helloworld/rest/RestHelloAction.java:63]: Credit to @dblock for the suggestion!

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:24]: > Just trying to understand why we need a separate `BaseExtensionRestHandler` class when `ExtensionRestHandler ` can handle the same? Can we move the code there?

Good question.  With JDK8 and `default` methods on interfaces, this could be moved up.  

It's generally my personal style preference to make a pretty sandwich:
 - Clean interface with no more than the class consumer needs (routes and a single `handleRequest()` method.
 - Messy default implementation that no sane user ever wants to deal with.
 - Clean implementation class that's really simple for end users writing the code.

In core, we have a `RestHandler` interface, which is really a functional interface as the only method is `handleRequest()` (and we should probably mimic that part?).

But we also have a `BaseRestHandler` which does the ugly guts of handling the request and only requires the end user to override `prepareRequest()`.

Would really like @dblock opinion on this one.

**Comment by mloufra** [comment]
> @mloufra FYI this can simplify your handlers in #213 and #221 if merged before they are.
> 
> A better example of the changes you might need is in https://github.com/opensearch-project/anomaly-detection/pull/726/files
> 
> However, it's a bit out of scope of what you're doing, so it's totally optional if you think it's too confusing.

Thanks @dbwiddis  for these information, I will take a look first.

**Comment by mloufra** [comment]
> 



> > @mloufra FYI this can simplify your handlers in #213 and #221 if merged before they are.
> > A better example of the changes you might need is in https://github.com/opensearch-project/anomaly-detection/pull/726/files
> > However, it's a bit out of scope of what you're doing, so it's totally optional if you think it's too confusing.

I will push the code change after this PR merged

**Comment by dbwiddis** [comment]
> I will push the code change after this PR merged

@mloufra this one may take a while, don't hold up your PR

**Comment by mloufra** [comment]
> > I will push the code change after this PR merged
> 
> @mloufra this one may take a while, don't hold up your PR

got it

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:24]: I'm fine with the new abstract class. @dblock if you anything to add

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:66]: Will named param has curly braces around? I thought it's already removed when passing the request

**Review by owaiskazi19 — DISMISSED** [review]

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:66]: The param is exactly what is in the `routes()` (or `routeHandlers()`) method, with the curly braces.   So for a route `GET` `/detectors/{detectorId}` the `handlerSplit` will be   `[ "detectors", "{detectorId}" ]`.  But the user submitting the request will have sent `GET` `/extensions/_anomaly_detection/detectors/ha7jrT78wK` so the `requestSplit` wlil be `[ "detectors", "ha7jrT78wK" ]`.

So we are splitting both strings by `/`, verifying they both have length 2, then checking each element. Index 0 is an exact string match, will match the first clause.  Index 1 is not an exact string match, but the handler starts and ends with curly braces so it passes, and the handler for this combination is filtered in.  

See the test here:
https://github.com/opensearch-project/opensearch-sdk-java/blob/69c8b392c0289285213067ed0d50954f67402a0e/src/test/java/org/opensearch/sdk/sample/helloworld/rest/TestRestHelloAction.java#L68

https://github.com/opensearch-project/opensearch-sdk-java/blob/69c8b392c0289285213067ed0d50954f67402a0e/src/test/java/org/opensearch/sdk/sample/helloworld/rest/TestRestHelloAction.java#L103


When the request is passed, it will have something

Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:66]: Thanks for the detailed information!

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java
Comment (dblock) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:24]: I am a Rubyist and thus generally avoid Base* classes because they are a Java thing. I think it's entirely a matter of preference.

**Review by owaiskazi19 — APPROVED** [review]

**Review by ryanbogan — APPROVED** [review]

**Pull request merged**
